discovery_client.go

```diff
@@ -0,0 +1,134 @@
+package rendezvous
+
+import (
+	"context"
+	"github.com/libp2p/go-libp2p-core/discovery"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"math"
+	"math/rand"
+	"sync"
+	"time"
+)
+
+type rendezvousDiscoveryClient struct {
+	rp        RendezvousPoint
+	peerCache sync.Map //is a map[string]discoveredPeerCache
+	rng       *rand.Rand
+	rngMux    sync.Mutex
+}
+
+type discoveredPeerCache struct {
+	cachedRegs []Registration
+	mux        sync.Mutex
+}
+
+func NewRendezvousDiscoveryClient(host host.Host, rendezvousPeer peer.ID) discovery.Discovery {
+	rp := NewRendezvousPoint(host, rendezvousPeer)
+	return &rendezvousDiscoveryClient{rp, sync.Map{}, rand.New(rand.NewSource(rand.Int63())), sync.Mutex{}}
+}
+
+func (c *rendezvousDiscoveryClient) Advertise(ctx context.Context, ns string, opts ...discovery.Option) (time.Duration, error) {
+	// Get options
+	var options discovery.Options
+	err := options.Apply(opts...)
+	if err != nil {
+		return 0, err
+	}
+
+	ttl := options.Ttl
+	var ttlSeconds int
+
+	// Default is minimum duration
+	if ttl == 0 {
+		ttlSeconds = 120
+	} else {
+		ttlSeconds = int(math.Round(ttl.Seconds()))
+	}
+
+	if err := c.rp.Register(ctx, ns, ttlSeconds); err != nil {
```

Comment:
I think we have changed the protocol to return the actual TTL, let's return it from the  `RendezvousPoint` interface and use that.

Reply:
I'm not sure I understand what you want to be done here. The `Register` functions in `RendezvousPoint` do not return TTL, I assume because the one implementation of `RendezvousPoint` doesn't modify the input TTL at all.

Would you like me to modify the `RendezvousPoint` and `RendezvousClient` interfaces to return the TTL? If so would you like it as an `int`, or `time.Duration`?

Reply:
Well, the returned TTL was added _after_ the rendezvous point interface was written.
Let's change the interface to return it.

Reply:
Changed the interface, I didn't get your feedback on `int` vs `time.Duration` so I went with `time.Duration` to be more compatible with the discovery interface